Add Collaborative WND (CWND) (#3529)
owncloud/docs: modules/admin_manual/pages/enterprise/external_storage/windows-network-drive_configuration.adoc

Compared to standard SMB access, WND has advanced features like:

. User lockout prevention and password reset
. More authentication mechanisms against the backend
. Listen to change information triggered by the backend
. Enhanced ACL support

> . Collaborative WND (CWND)

Reviewer: There is also collaborative SMB, so I'm not sure if we want to mention the "CWND" as an advantage

Author: Yes, because it is specific to WND.
Even there is a similar functionality in SMB, SMB has eg no listener ect and there is the need for differenciation.
The upcoming CSMB description will then have only a subset of the CWND text

.Brief Description of Advanced Features:

User lockout prevention and password reset::
Depending on the Windows or Samba policy, the user could get locked out of his account if he enters a wrong password a number of times. The lockout prevention tries to avoid this from happening, by resetting the password if it is wrong. For the case of ownCloud's standard SMB connector, the password won't be reset. It could happen, that the user gets locked out of the file server.

More authentication mechanisms against the backend::
Please see the details about the xref:enterprise/external_storage/enterprise_only_auth.adoc[Enterprise-Only Authentication Options]

Listen to change information triggered by the backend::

> Native Windows File Servers provide the ability to proactively send change notifications made to files and folders somewhere in a share. This mechanism then updates the ownCloud database and provides the changes made to accessing users. Users do not need to manually check for changes in all possible locations of their mount. Changes processed are also propagated to sync clients automatically.

Reviewer: I think samba can do that as long as all the accesses are done through the SMB protocol. This is different than windows, where you get notifications even if the changes are done directly in the storage.
Taking into account the requirements, we might want to skip this information from here. It won't be brief otherwise.

Author: Do a text change suggestion to satisfy both native and samba file servers. This would help a lot.

Reviewer: Maybe adding the information as a note is better. The requirements are simply different

Author: We have in all items a description of what it is about. We cant just drop a "note" here. We need a proper explaining text. Suggestions are really welcomed.

Reviewer: "Samba can send file notifications as well, as long as all the file actions are performed through the SMB protocol. However, it won't work if the action is performed directly inside the FS used by Samba"

Author: fixed

Enhanced ACL support::

> With enhanced ACL support, both SMB and WND evaluate the file attributes (whether the file is hidden or read-only) to decide what ownCloud permissions should the file or folder have in ownCloud. On top of this, WND can also evaluate the ACLs by using the `ocLdapPermissionManager` in the mount point configuration. This will bring more accurate permissions to ownCloud, specially if each user could have different permissions over the files in Windows.

Reviewer: It might be important to warn that this ACL support doesn't apply to the collaborative WND (at least for now)

Author: 👍

Author: Can you tell more about. Is it just not working? or ist it breaking something? What is the effect the user / admin is confronted with?

Reviewer: For listing, we're using the service account, which could have specific permissions. Since we're using the SMB file attributes by default, unless the file is read-only, it will have all permission. Those permissions are stored in the filecache.

Problem 1: if we evaluate ACLs for the service account, the ownCloud permissions will be limited to read-only (unless the service account have permissions to write or full permissions in the SMB server).

Problem 2: the permissions are cached, but they aren't re-evaluated unless the file changes. For the same fileid, each user could have different permissions. If userA has only read access, we might cache that the permissions for the fileid are read-only. However, when userB access, since the permissions are cache, he'll also have read-only access (in ownCloud) even though he could have full access in the SMB server.

For now, ownCloud is being permissive by evaluating only the file attribute, so usually, ownCloud will allow anything to any user. In any case, the permissions are being enforced by the SMB server, so if userA cannot write a file an error will be reported.

Author: As discussed, I added a text at the end of this description to clarify the mandatory use of the default nullperm manager when using cwnd.

Collaborative WND (CWND)::

> Compared to a standard WND mountpoint, a collaborative WND mount offers enhanced features. In a CWND, each user shares the same ownCloud internal informations for files and folders based on its internal identification (file_id). This means that _comments_ and _tags_ can be shared to all users accessing files and folders of this mount without the need that users muste be member of the mount from ownCloud point of view. A CWND can only be set by an admin in menu:Settings[Admin > Storage] but not in the users section. CWND takes the benefit like you share a mount and can therefore share eg comments and tags, but removes the drawback that you have only one accessing user, the sharer. With CWND, each accessing user has its own access with his credentials but shares additional information with other users accessing the same data. See the table below to compare the differences.

Reviewer 2: users *muste* -> users *must*

+
.Collaborative WND Differences Based on the Mount Type
[cols=".^15%,.^35%,.^35%",options="header",caption=]
|===
|
^| Windows Network Drive
^| Windows Network Drive (collaborative)

h| Login Credentials

> a| * User Credentials +

Author: at login it is correct having capital C ?

> a| * User Credentials +

Reviewer 3: Yes, because it's like a headline in the table.

. Enhanced ACL support
. Collaborative WND (CWND)

.Brief Description of Advanced Features:

User lockout prevention and password reset::
Depending on the Windows or Samba policy, the user could get locked out of his account if he enters a wrong password a number of times. The lockout prevention tries to avoid this from happening, by resetting the password if it is wrong. For the case of ownCloud's standard SMB connector, the password won't be reset. It could happen, that the user gets locked out of the file server.

More authentication mechanisms against the backend::
Please see the details about the xref:enterprise/external_storage/enterprise_only_auth.adoc[Enterprise-Only Authentication Options]

Listen to change information triggered by the backend::
Native Windows File Servers provide the ability to proactively send change notifications made to files and folders somewhere in a share. This mechanism then updates the ownCloud database and provides the changes made to accessing users. Users do not need to manually check for changes in all possible locations of their mount. Changes processed are also propagated to sync clients automatically.

Enhanced ACL support::

> With enhanced ACL support, both SMB and WND evaluate the file attributes (whether the file is hidden or read-only) to decide what ownCloud permissions should the file or folder have in ownCloud. On top of this, WND can also evaluate the ACLs by using the `ocLdapPermissionManager` in the mount point configuration. This will bring more accurate permissions to ownCloud, specially if each user could have different permissions over the files in Windows.

Reviewer 4: ```suggestion
With enhanced ACL support, both SMB and WND evaluate the file attributes (whether the file is hidden or read-only) to decide what ownCloud permissions the file or folder should have in ownCloud. On top of this, WND can also evaluate the ACLs by using the `ocLdapPermissionManager` in the mount point configuration. This will bring more accurate permissions to ownCloud, specially if each user could have different permissions over the files in Windows.
```

> With enhanced ACL support, both SMB and WND evaluate the file attributes (whether the file is hidden or read-only) to decide what ownCloud permissions the file or folder should have in ownCloud. On top of this, WND can also evaluate the ACLs by using the `ocLdapPermissionManager` in the mount point configuration. This will bring more accurate permissions to ownCloud, specially if each user could have different permissions over the files in Windows.

Reviewer 3: ```suggestion
With enhanced ACL support, both SMB and WND evaluate the file attributes (whether the file is hidden or read-only) to decide what ownCloud permissions the file or folder should have in ownCloud. On top of this, WND can also evaluate the ACLs by using the `ocLdapPermissionManager` in the mount point configuration. This will bring more accurate permissions to ownCloud, especially when each user can have different permissions for the files in Windows.
```

Collaborative WND (CWND)::

> Compared to a standard WND mountpoint, a collaborative WND mount offers enhanced features. In a CWND, each user shares the same ownCloud internal informations for files and folders based on its internal identification (file_id). This means that _comments_ and _tags_ can be shared with all users accessing files and folders of this mount without the need that users must be member of the mount from ownCloud point of view. A CWND can only be set by an admin in menu:Settings[Admin > Storage] but not in the users section. With CWND, you lose the benefit of sharing a mount and therefore sharing e.g. comments and tags, but it removes the drawback that you only have one accessing user, the sharer. With CWND, all accessing users have their own access with their credentials but share additional information with other users accessing the same data. See the table below to compare the differences.

Reviewer 4: ```suggestion
Compared to a standard WND mountpoint, a collaborative WND mount offers enhanced features. In a CWND, each user shares the same ownCloud internal information for files and folders based on its internal identification (file_id). This means that _comments_ and _tags_ can be shared with all users accessing files and folders of this mount without the need that users must be members of the mount from an ownCloud point of view. A CWND can only be set by an admin in menu:Settings[Admin > Storage] but not in the users section. With CWND, you lose the benefit of sharing a mount and therefore sharing e.g. comments and tags, but it removes the drawback that you only have one accessing user, the sharer. With CWND, all accessing users have their own access with their credentials but share additional information with other users accessing the same data. See the table below to compare the differences.
```

Reviewer 4: "This means that _comments_ and _tags_ can be shared with all users" and "With CWND, you lose the benefit of sharing a mount and therefore sharing e.g. comments and tags"
I am confused - are comments and tags shared or not? In what situations?

Author: You are right, needs rewriting.
I need an info from @jvillafanez , he got the question already. When I get it, I will fix this.

Reviewer 5: This part completely confuses me: "With CWND, you lose the benefit of sharing a mount" -- I thought the collaborative feature was about improving handling of shared resources. (but forgive me, I haven't read the rest yet.)
The explanation that follows introduces more topics: We learn that "only .. one accessing user" is a drawback. That seems to contradict the introduction of one service user to rule them all -- which should be more a solution than a problem.

Author: text is fixed

+
.Collaborative WND Differences Based on the Mount Type
[cols=".^15%,.^35%,.^35%",options="header",caption=]
|===
|
^| Windows Network Drive
^| Windows Network Drive (collaborative)

> h| Login Credentials

Author: ```suggestion
h| Login credentials
```

a| * User credentials +
* Credentials of the sharer
a| * `Log-in credentials, saved in session` +
* `Log-in credentials, saved in database` +
* `User entered, stored in database`

h| File ID
| Unique per user or from the sharer
| Same for all users accessing this mount

h| Access Rights
| From the accessing user or the sharer
| From the accessing user

h| Activities +
Comments +
Tags
a| * No shared access +
** Visibility limited to the user +
* Shared access +
** Comments and tags are shared, access based on the sharer
| Comments and tags are shared, access based on the user
|===

.More WND Properties
Mounts to a Windows or Samba file server are labeled with a little four-pane Windows-style icon, and the left pane of your Files page includes a Windows Network Drive filter. Figure 1 shows a new Windows Network Drive share marked with a red warning which indicates that ownCloud cannot connect to the share. The reason is that it may require the user to login, or it is not available, or there is an error in the configuration.

> Files are synchronized bi-directionally, and you can create, upload, and delete files and folders.

Reviewer 3: ```suggestion
Files are synchronized bidirectionally, and you can create, upload and delete files and folders.
```

ownCloud server admins can create Windows Network Drive mounts and optionally allow users to set up their
own personal Windows Network Drive mounts.

Depending on the authentication method, passwords for each mount are encrypted and stored in the ownCloud
database, using a long random secret key stored in `config.php`. This allows ownCloud to access the shares

> when the users who own the mounts are not logged in. This access will not work if the mount is session based, where passwords are not stored and available only for the current active session.

Reviewer 4: ```suggestion
when the users who own the mounts are not logged in. This access will not work if the mount is session based, where passwords are not stored and are available only for the current active session.
```

=== Creating a New Share

.When you create a new WND share, you need three things:

* the login credentials for the share,

> * the server address, the share name and
> * The folder you want to connect to

Author: ```suggestion
* the server address, the share name and
* the folder you want to connect to
```

=== Creating a New Share

.When you create a new WND share, you need three things:

* the login credentials for the share,
* the server address, the share name and

> * the folder you want to connect to

Reviewer 3: ```suggestion
* the folder you want to connect to.
```

=== Creating a New Share

.When you create a new WND share, you need three things:

* The login credentials for the share

> * The server address, the share name; and

Author: will check

NOTE: `wnd.activity.sendToSharees` key depends on the `wnd.activity.registerExtension` key to take effect.

=== Collaborative WND

> CWND can only be set by an admin in menu:Settings[Admin > Storage]. This mount type cannot be selected by users in the user section. To prepare access for your mount point using the CWND mount type, you must provide a _Service Account_ (SA) which is an oridnary SMB user granting read access to the share you want to mount. You can use one SA for all CWND mounts or separate ones. The SA is used to gather the contents of a share, used by the WND Listener and provides a common `file_id` to all accessing users, while the accessing user has only access to those files and folders he is granted for.

Reviewer: "which is an ordinary SMB user"

Author: How to describe/name a user which is created in AD or in Samba which can access file services but who is not created in ownCloud?

Reviewer: I think that's fine. This was just for the typo :)

. As an admin, goto menu:Settings[Admin > Storage] and create a new CWND based mount point.
+
.Figure 5. Add a Collaborative Windows Network Drive Mount
image:enterprise/external_storage/windows_network_drive/cwnd_add_storage.png[Add Collaborative Windows Network Drive Mount, width=50%]
. Chose any name for the mount point that fits your needs.
. Select User Login Type.
+
[IMPORTANT]
====

> Only the following three are valid selections for CWND

Reviewer: Technically, all of them are valid, but it's true that some of them don't have sense to be "collaborative".
Both global credentials and username+password don't have sense to be "collaborative". The credentials hardcoded in config file can also fit as "valid" here, although the use case is pretty limited from my point of view, so maybe we can skip it too.

Author: What about to rephrase this line to highlight that those are working AND sensful ones, any suggestions?

Reviewer: Yes, that's fine

Author: Suggestion see: https://github.com/owncloud/docs/pull/3529#discussion_r627472206

> Only the following three are valid selections for CWND:

Author: ```suggestion
The following three are sensful and working selections for CWND:
```

> The following three are sensful and working selections for CWND:

Author: ```suggestion
The following three are sensible and working selections for CWND:
```

[loweralpha]
. `Log-in credentials, saved in session`
. `Log-in credentials, saved in database`
. `User entered, stored in database` ^[1]^
+

> [1] Must be used in case user authentication is made with OIDC

Reviewer 4: ```suggestion
[1] Must be used if user authentication is made with OIDC
```

====
+
.Figure 6. Select How User Logs Into the Mount Point
image:enterprise/external_storage/windows_network_drive/cwnd_login_possibilities.png[Select How User Logs Into the Mount Point]
[loweralpha]
.. `Log-in credentials, saved in session`
+

> When the user logs into ownCloud via browser, the credentials to authenticate CWND are taken from this login. These credentials immediate end when the user logges out because the session has ended. This may impact client synchronisation or any access to other users via sharing. In case, no synchroinisation can be processed and sharees will lose access tho their shared resource when the sharer logges out. _This login type is by design not compatible with OIDC authentication._

Reviewer: "no synchronisation can be processed and sharees will lose access to"

Reviewer: > In case, no synchroinisation can be processed and sharees will lose access tho their shared resource when the sharer logges out

It seems to suggest that the access is possible while the user is logged in, but this isn't true. Even if userA is logged in, neither userB nor ownCloud will be able to access to userA's credentials, so nobody can access to the storage on behalf of userA (shares, file scan, etc)

Author: If I am userA and share my mount (or a subfolder of my mount) with userB,
userB accesses the mount with userAs credentials - this is my understanding - am I wrong?

Reviewer: That's right. However, userB won't be able to access to userA's credentials in this case even if userA is logged in because userB cannot access to the session information of userA

Author: Waiting for input (see chat)

Author: got the input, working on it

Author: fixed

.. `Log-in credentials, saved in database`
+
Similar to `Log-in credentials, saved in session`, the credentials to authenticate CWND are taken from the login but saved into the ownCloud database. Any relogin also updates the database entry. As the credentials to access CWND are taken from the database, a user logout will not stop CWND access and serving data is continued. _This login type is by design not compatible with OIDC authentication._
.. `User entered, stored in database`
+

> User login to ownCloud and providing credentials to access the CWND mount are completely separated. The user may see post login to ownCloud his CWND mounts being inaccessible. To regain access, a pop up will appear when clicking on the mount point and the user must enter his share credentials which are then stored into the ownCloud database. As the credentials to access CWND are taken from the database, a user logout will not stop CWND access and serving data is continued. _This login type is by design *the only compatible with OIDC authentication*._

Reviewer: We might want to skip the part about the popup. I don't think it will work properly with CWND.
I think the complain will come from a wrong service account password (usually), but the credentials we'll update are the user's ones, which might be right.

Author: The SA pwd has not changed and is independent of the users pwd. The SA still can access the resource. But as userA eg. has not logged in before in oC, there are no credentials for this mount for userA present - the mount will turn unavailable for userA. userA must now enter the credentials. Q: how will userA provide the credentials to regain access ?

Reviewer: In the personal storage of the user, he should see a section to add the credentials for the mount. I think the admin needs to enable the personal storages though.

Author: THIS IS NEW - NEEDS CLARIFICATION
@jnweiger fyi

Reviewer: admin user:
![Screenshot from 2021-05-07 10-47-27](https://user-images.githubusercontent.com/1477829/117424125-c0e5cf80-af21-11eb-9fd9-54f26b174cb3.png)

target user:
![Screenshot from 2021-05-07 10-47-57](https://user-images.githubusercontent.com/1477829/117424167-ca6f3780-af21-11eb-97b9-c4918f3bf7b0.png)

The user would need to fill the credentials for the mount point. He can use any valid SMB account.

Author: fixed

+
. Configure this mount point by adding data into the corresponding fields
+
The password for the Service Account can be temporarily any value to satisfy the creation of the mount point. The functional password is entered in the next step.
+
.Figure 7. Enter Connection Info and the Service Account
image:enterprise/external_storage/windows_network_drive/cwnd_fields.png[Enter Connection Info and Service Account]
+
When everything has been entered correct, the mount point gets a green button on the left.
. Enter the Password of the Service Account
+
First get the <mount id> of the newly created mount point by running the command below in a shell:
+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} files_external:list --short
----
+

> This will procuce an output like the following:

Reviewer 3: ```suggestion
This will produce an output like the following:
```

+
----
+----------+-------------+------------------+-------------------+------+-------+
| Mount ID | Mount Point | Applicable Users | Applicable Groups | Auth | Type  |
+----------+-------------+------------------+-------------------+------+-------+
| 1        | /SFTP       | All              |                   | User | Admin |
| 2        | /CWND       | All              |                   | User | Admin |
+----------+-------------+------------------+-------------------+------+-------+
----
+

> Use the <mount-id> number of your fresh created CWND mount for the next command:

Reviewer 3: ```suggestion
Use the <mount-id> number of your freshly created CWND mount for the next command:
```

+
[source,console,subs="attributes+"]
----
{occ-command-example-prefix} wnd:set-service-account <mount-id>
----
+

> This command will ask you for the password of the used Service Account and stores it encyrpted in the database.

Reviewer: "stores it encrypted"

> This command will ask you for the password of the used Service Account and stores it encyrypted in the database.

Reviewer 4: ```suggestion
This command will ask you for the password of the used Service Account and stores it encrypted in the database.
```

+

> NOTE: In case the password for the SA changes, you have to redo this step for each CWND mount point.

Reviewer 4: ```suggestion
NOTE: If the password for the SA changes, you have to redo this step for each CWND mount point.
```

. As an admin, go to menu:Settings[Admin > Storage] and create a new CWND based mount point.
+
.Figure 5. Add a Collaborative Windows Network Drive Mount
image:enterprise/external_storage/windows_network_drive/cwnd_add_storage.png[Add Collaborative Windows Network Drive Mount, width=50%]
. Chose any name for the mount point that fits your needs.
. Select user login type.
+
[IMPORTANT]
====
The following three are sensible and working selections for CWND:
[loweralpha]
. `Log-in credentials, saved in session`
. `Log-in credentials, saved in database`
. `User entered, stored in database` ^[1]^
+
[1] Must be used if user authentication is made with OIDC
====
+

> .Figure 6. Select How User Logs Into the Mount Point

Reviewer 3: ```suggestion
.Figure 6. Select How User Logs in to the Mount Point
```

image:enterprise/external_storage/windows_network_drive/cwnd_login_possibilities.png[Select How User Logs Into the Mount Point]
[loweralpha]
.. `Log-in credentials, saved in session`
+
When the user logs in to ownCloud via a browser, the credentials to authenticate CWND are taken from this login. These credentials immediately end when the user logs out because the session has ended.
+
* _This login type can not be set to `Enable Sharing`._
* _This login type is by design not compatible with OIDC authentication._
.. `Log-in credentials, saved in database`
+

> Similar to `Log-in credentials, saved in session`, the credentials to authenticate CWND are taken from the login but saved in the ownCloud database. Any re-login also updates the database entry. As the credentials to access CWND are taken from the database, a user logout will not stop CWND access and serving data is continued like for for synchronisation.

Reviewer 3: ```suggestion
Similar to `Log-in credentials, saved in session`, the credentials to authenticate CWND are taken from the login but saved in the ownCloud database. Any re-login also updates the database entry. As the credentials to access CWND are taken from the database, a user logout will not stop CWND access and serving data is continued, e.g. for synchronization.
```

+
* _This login type can be set to `Enable Sharing`._
* _This login type is by design not compatible with OIDC authentication._
.. `User entered, stored in database`
+

> User login to ownCloud and providing credentials to access the CWND mount are completely separated. After logging in to ownCloud, the user may see his CWND mounts marked inaccessible. To regain access, the user must enter his share credentials in menu:Settings[Personal > Storage] which are then stored into the ownCloud database. As the credentials to access CWND are taken from the database, a user logout will not stop CWND access and serving data is continued like for for synchronisation.

Reviewer 3: ```suggestion
User login to ownCloud and providing credentials to access the CWND mount are completely separated. After logging in to ownCloud, the user may see his CWND mounts marked inaccessible. To regain access, the user must enter his share credentials in menu:Settings[Personal > Storage] which are then stored into the ownCloud database. As the credentials to access CWND are taken from the database, a user logout will not stop CWND access and serving data is continued, e.g. for synchronization.
```
